### PR TITLE
Fix disabled files app

### LINKED ISSUE #13840: Files app disabled / config defaults issue ?
I upgraded my openSUSE 13.1 package of owncloud-7.0.4-17 to owncloud-8.0.0~beta2.
Then I ran `sudo -uwww ./occ upgrade`, all went fine.

Now WebDAV doesn't work and the log says:
```
{"reqId":"187e1bbd19a92b4d7a45775be28ab0eb","remoteAddr":"85.180.4.155","app":"remote","message":"App not installed: files","level":4,"time":"2015-02-02T21:13:54+00:00","method":"PROPFIND","url":"\/remote.php\/webdav\/"}
{"reqId":"4f1045ced87c86d8bf1388425fa1bd1d","remoteAddr":"85.180.4.155","app":"remote","message":"App not installed: files","level":4,"time":"2015-02-02T21:14:26+00:00","method":"PROPFIND","url":"\/remote.php\/webdav\/"}
{"reqId":"c056c3692650b029d06b1217fb98f8ff","remoteAddr":"85.180.4.155","app":"remote","message":"App not installed: files","level":4,"time":"2015-02-02T21:14:58+00:00","method":"PROPFIND","url":"\/remote.php\/webdav\/"}
{"reqId":"80e7624c491895e35032a46d468f4f4b","remoteAddr":"85.180.4.155","app":"remote","message":"App not installed: files","level":4,"time":"2015-02-02T21:15:30+00:00","method":"PROPFIND","url":"\/remote.php\/webdav\/"}
```

Checking the DB:
```
MariaDB [owncloud]> select * from oc_appconfig where appid='files';
+-------+-----------------------------------+-------------+
| appid | configkey                         | configvalue |
+-------+-----------------------------------+-------------+
| files | backgroundwatcher_previous_file   | 16874       |
| files | backgroundwatcher_previous_folder | 13963       |
| files | default_quota                     | 1 GB        |
| files | installed_version                 | 1.1.9       |
| files | types                             | filesystem  |
+-------+-----------------------------------+-------------+
```

There is no "enabled" key at all ?!

I checked my backup from 7.0.4 and the key was *also* missing.

So I guess that in 7.0.4 the default value for "enabled" was "yes" and now in 8.0.0 we changed it to "no" ? Was this on purpose ? Or is it a side-effect of the recent refactorings of the app config stuff ?

Note: this is different than https://github.com/owncloud/core/issues/13523 where the disabling was done on purpose, and also has the "enabled" key present.

I believe this is a different issue here, related to default config values handling.

Note that I upgraded since OC 4 so maybe that key never existed, so maybe people with similar setups will experience it as well.

@MorrisJobke @LukasReschke

### PR DESCRIPTION
**Pull request by MorrisJobke**
cc @LukasReschke @PVince81

fixes #13840

### CODE REVIEW

**Comment by MorrisJobke** [comment]
This fix contains proper handling of the attempt to disable the files app -> exception.

The second commit contains a repair step that fixes the state of a disabled files app.

**Comment by MorrisJobke** [comment]
> The inspection completed: 6 new issues, 7 updated code elements

Just new ones, because others are fixed and scrutinizer didn't hit the limit anymore ;)

**Comment by DeepDiver1975** [comment]
> The second commit contains a repair step that fixes the state of a disabled files app.

is that one really necessary? Handling the app's state pure in code should have no impact on the state in the db. I'd prefer if we can get rid of this - the less code to maintain - the better. THX

**Comment by PVince81** [comment]
@DeepDiver1975 I think I rather prefer to not have special cases hard-coded in core like "force files app enabling even if disabled". I could think of use cases where the files app + webdav could be disabled for setups that only use calendar/contacts (might not work now but maybe in the future).
Apps like notes should still be able to write files, but these won't be visible/accessible directly for the user.

So this is basically the concern of maintaining a repair step vs maintaining the hard-coded exception case.

**Comment by DeepDiver1975** [comment]
I fully agree - all these special case handling is killing us - but honestly speaking: why enable something with an repair step which is not supposed to be disabled.

As of today files is not allowed to be disabled - so let's ensure this.

> I could think of use cases where the files app + webdav could be disabled for setups that only use calendar/contacts (might not work now but maybe in the future).

Some day - maybe ...

**Comment by PVince81** [comment]
Well there's the point of view of data consistency. If the DB says it's enabled, then it is.
But anyway, I don't care too much about which solution with go for, just wanted to mention that use case :smile:

**Comment by MorrisJobke** [comment]
The problem here is that there where cases in the wild, where the files app wasn't enabled. So I ensured that we currently doesn't run into this situation again.(exceptions) Afterwards I added code to migrate from this broken code. We can drop this repair step once 8.0 is out. But I don't want to risk that some instances have this broken file app because it was disabled at some point in time where it was possible and had no impact due to the hardcoded enabled state. And the exceptions can be dropped once we support files app to be disabled.

**Comment by PVince81** [comment]
#### Tests
- [x] delete "enabled" flag in OC 7 then upgrade
- [ ] delete "enabled" flag in OC 8, connect with WebDAV :no_entry_sign: FAIL
- [ ] manually disable "files" app

File: lib/private/app.php [hunk 1]
```diff
@@ -321,6 +321,9 @@ public static function removeApp($app) {
 	 * @param string $app app
 	 */
 	public static function disable($app) {
+		if($app === 'files') {
+			throw new \Exception("App 'files' can't be disabled.");
```
Comment (nickvergessen) [inline lib/private/app.php]: The other messages don't wrap the app name with `'`
Should we keep that consistent for now?

```
nickv@nickv-lt-mint ~/ownCloud/Master $ ./occ app:disable activity
activity disabled
nickv@nickv-lt-mint ~/ownCloud/Master $ ./occ app:disable activity
activity enabled
nickv@nickv-lt-mint ~/ownCloud/Master $ ./occ app:disable files
App 'files' can't be disabled.
nickv@nickv-lt-mint ~/ownCloud/Master $ ./occ app:enable files
files is already enabled
```

**Comment by DeepDiver1975** [comment]
> delete "enabled" flag in OC 8, connect with WebDAV :no_entry_sign: FAIL

and this is honestly stupid - disabling files is not allowed (yet) - please make sure the app is enabled at any means.

**Comment by MorrisJobke** [comment]
I did following tests:

* files app was disabled (either by `enabled == no` or by a missing `enabled` entry in the DB) in pre 8.0
* try to disable the files app with the command line (because it isn't listed in the apps management)

There is no other valid way to disable the app. Otherwise you broke it by intention and that is something we won't support.

File: lib/repair/enablefilesapp.php [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+use OC\Hooks\BasicEmitter;
+use OC\RepairStep;
+use OCP\IConfig;
+
+/**
+ * Class EnableFilesApp - enables files app if disabled
+ *
+ * @package OC\Repair
+ */
+class EnableFilesApp extends BasicEmitter implements RepairStep {
```
Comment (LukasReschke) [inline lib/repair/enablefilesapp.php]: Add TODO and/or FIXME to state that we can remove this with ownCloud 9?

File: lib/private/app.php (hunk 1 above)
Comment (DeepDiver1975) [inline lib/private/app.php]: yes - should be

````
files cannot be disabled.

Comment (MorrisJobke) [inline lib/private/app.php]: Done

File: lib/repair/enablefilesapp.php (hunk 2 above)
Comment (MorrisJobke) [inline lib/repair/enablefilesapp.php]: Done

**Comment by MorrisJobke** [comment]
@owncloud-bot retest this please

**Comment by Raydiation** [comment]
Hm, running php -f console.php app:disable files throws gives me this exception:

PHP Warning:  flock() expects parameter 1 to be resource, boolean given in /srv/http/owncloud/lib/private/config.php on line 178
```
Warning: flock() expects parameter 1 to be resource, boolean given in /srv/http/owncloud/lib/private/config.php on line 178
An unhandled exception has been thrown:
exception 'Exception' with message 'Could not acquire a shared lock on the config file /srv/http/owncloud/config/config.php' in /srv/http/owncloud/lib/private/config.php:179
Stack trace:
#0 /srv/http/owncloud/lib/private/config.php(38): OC\Config->readData()
#1 /srv/http/owncloud/lib/base.php(100): OC\Config->__construct('/srv/http/owncl...')
#2 /srv/http/owncloud/lib/base.php(473): OC::initPaths()
#3 /srv/http/owncloud/lib/base.php(1007): OC::init()
#4 /srv/http/owncloud/console.php(12): require_once('/srv/http/owncl...')
```

@LukasReschke ?

**Comment by LukasReschke** [comment]
Different issue. File a new bug please.

**Comment by Raydiation** [comment]
Ok, @LukasReschke my config file was not readable as normal user :) 

Can confirm that this is working, you cant disable the files app over the commandline, code looks good :+1:

**Comment by DeepDiver1975** [comment]
okya - let's move on :+1:

**Comment by DeepDiver1975** [comment]
THX

**Pull request merged**